rclpy/rclpy/context.py

```diff
@@ -127,7 +127,7 @@ def on_shutdown(self, callback: Callable[[], None]):
                 if ismethod(callback):
                     self._callbacks.append(weakref.WeakMethod(callback, self._remove_callback))
                 else:
-                    self._callbacks.append(weakref.ref(callback, self._remove_callback))
+                    self._callbacks.append(callback)
```

Comment:
> I could also see the case of always retaining shared ownership and the user explictily passing a weak object when they need that, but that would be an unexpected change (particularly for methods).

Oh hmm. This is interesting. I don't see a better way, and it does seem like the behavior would be intuitive.